pilot/pkg/xds/ads.go

```diff
@@ -340,16 +371,26 @@ func (s *DiscoveryServer) shouldRespond(con *Connection, request *discovery.Disc
 	con.proxy.WatchedResources[request.TypeUrl].LastRequest = request
 	con.proxy.Unlock()
 
+	// Signal waiting semaphore in push operations that a valid
+	// ACK has been processed as determined by matched nonce
+	if features.EnableFlowControl {
+		select {
+		case <-con.semaphore:
```

Comment:
This should be moved in front. To cover the case of NACK

Reply:
do you mean the exit thread case should be moved in front? I don't quite understand the comment.

Reply:
You mean the front of the function? I'm not sure that would work - but can give it a try. The location currently represents when the nonce matches.

Cheers,
-steve